### PR TITLE
fix for short address attack

### PR DESCRIPTION
**Pull request by jdetychey**
as suggested by /u/izqui9 here https://www.reddit.com/r/ethereum/comments/63s917/worrysome_bug_exploit_with_erc20_token/dfwmhc3/
Attack description: 
https://blog.golemproject.net/how-to-find-10m-by-just-reading-blockchain-6ae9d39fcd95

### CODE REVIEW

**Comment by miohtama** [comment]
I read Golem blog post https://blog.golemproject.net/how-to-find-10m-by-just-reading-blockchain-6ae9d39fcd95 and I am trying to understand better what happened. Now there is a confirmation that this was a bug specific to the exchange implementation. 

Though I see this as a risk in the future, I am not sure if this should be fixed like this. Rationale here is that the bug was not specific to a `transfer` function, but as far as I understand from the post, the exchange code would have encoded data any function call invalidly. Thus, one would have to make extra effort to protect all functions against invalid data payloads.

**Comment by maraoz** [comment]
Personally I'm not sure if this should be fixed at the smart contract level. [The attack](https://blog.golemproject.net/how-to-find-10m-by-just-reading-blockchain-6ae9d39fcd95) seems like an implementation bug in how exchanges handle user inputs and craft transactions with them. If exchanges use standard tools and good security practices this shouldn't be needed. 

That said, this seems like a low-cost protection that would add to the overall security of tokens. As such, it would be a step towards increasing [in-depth security](https://github.com/OpenZeppelin/zeppelin-solidity/blob/master/CONTRIBUTING.md#d0---security-in-depth) of OpenZeppelin contracts. 

Let's have some community discussion before we merge this please. I'm not against but still not convinced.

**Comment by makoto** [comment]
Should have test coverage. Also CI test is failing...

**Comment by simondlr** [comment]
It's a useful addition, but I would agree: is the contract responsible for this security? It could be, and so your consideration arrives at cost of error vs cost of gas.

In the event of an error, you would still have to consider if all OpenZeppelin tokens over the course of its deployment won't add up to the cost of that error?

So you'll have to probably do some probability weights to determine if it's worth it. What's the likelihood of this attack being exploited and what's the cost associated with it (risk of event occurring * cost), then see if that adds to gas cost over a certain period of time.

**Comment by makoto** [comment]
Is the extra gas cost something we can quantify?

**Comment by simondlr** [comment]
> Is the extra gas cost something we can quantify?

Yes. The cost of the modifier is additional gas cost vs not having the protection.

This gas cost for all Zeppelin tokens need to be added up and compared to the cost of losing value due to the protection not being there. Depends on how you weight the probability of these errors occurring, and what the expected losses will be.

**Comment by makoto** [comment]
@simondlr I mean the actual gas cost. Is it always x wei (if so how much is x?) or variable depending on other factors?

**Comment by simondlr** [comment]
> Is it always x wei (if so how much is x?) or variable depending on other factors?

I assume so, yes. Essentially the cost of the extra code being deployed + the modifier being called for every transfer.

This line seems like a fixed cost.
> assert(msg.data.length == size + 4);

**Comment by maurelian** [comment]
Wouldn't the cost be "x gas", rather than "x wei"?

**Comment by makoto** [comment]
Looks like the discussion is already going on at Solidity level. https://github.com/ethereum/solidity/issues/2109

Probably duplicate effort?

**Comment by izqui** [comment]
I think this 'more gas for more security' discussion is very interesting and goes beyond this specific issue. 

After reading this https://github.com/ethereum/solidity/issues/510, my personal opinion is that Zeppelin should have a 'StandardToken' and a 'UnsafeStandardToken'. The unsafe token would be cheaper to run for use cases where the gas difference is really a game changer, this implementation could lack also security measures such as using safe math.

I think the industry standard should be to add as much security as possible (even if some things look like an overkill and by having to pay the gas every time) to the contracts to prevent against hypothetic multi-million dollar looses. Once the contracts can pay for their own gas, this would be a decision up to the developer/creator to individually asses the risk and decide how much security she wants and wants to pay for.

**Comment by simondlr** [comment]
@izqui I like that approach.

The "standard" being the most secure contract, and then creating alternatives for developers that know what they are doing (substituting some security features for gas savings, for example).

**Comment by maraoz** [comment]
@izqui @simondlr sounds good.
PR still failing for some reason. Will restart travis build and see what happens

**Comment by izqui** [comment]
Oh, it is the vested token test that is failing. The `grantVestedTokens(...)` function calls the transfer function under the hood and it is doing the assertion that the payload is 68 bytes, but the payload for `grantVestedTokens` is bigger.

Modifying the assertion to `assert(msg.data.length >= size + 4)` should fix this and the assertion would still work, as the problem is with smaller not with bigger payloads (where Solidity will throw for us, I think).

The name of the modifier should be changed to something like 'minimumPayloadSize'.

@jdetychey

**Comment by jdetychey** [comment]
done thx

**Comment by izqui** [comment]
LGTM!

**Comment by maraoz** [comment]
LGTM

File: contracts/token/BasicToken.sol [hunk 1]
```diff
@@ -13,7 +13,15 @@ contract BasicToken is ERC20Basic, SafeMath {
 
   mapping(address => uint) balances;
 
-  function transfer(address _to, uint _value) {
+/*
+ * Fix for the ERC20 short address attack  
+ */
+  modifier onlyPayloadSize(uint size) {
+     assert(msg.data.length >= size + 4);
```
Comment (chriseth) [inline contracts/token/BasicToken.sol]: As you are checking a condition on inputs, shouldn't this rather be a `require()`?

Comment (izqui) [inline contracts/token/BasicToken.sol]: The code is using Solidity v0.4.8, the assert function is implemented in SafeMath.sol, it is not the solidity one

Whenever Zeppelin migrates to v0.4.10 i agree that should be a require.

Comment (chriseth) [inline contracts/token/BasicToken.sol]: Ah sorry and thanks for the clarification :-)

**Pull request merged**